Add support for oci-dir output (fixes #1865)

septatrix opened the pull request:
A first draft for rudimentary OCI image outputs

PS: I intend to squash this once more questions are answered/resolved

septatrix commented on mkosi/archive.py:
```diff
@@ -63,6 +66,72 @@ def make_tar(src: Path, dst: Path, *, tools: Path = Path("/"), sandbox: Sequence
+
+    layer_archive = context.staging / context.config.output_tar
+    with layer_archive.open("rb") as f:
+        layer_digest = hashlib.file_digest(f, "sha256").hexdigest()
```
`file_digest` only exists in Python 3.11, I will change this in the final version similar to a loop similar to what is done for the `SHA256SUM` logic

septatrix replied:
We could also backport this but I am not sure how to reference licenses in such cases, instead I moved the existing `hash_file` into utils and also use it here

septatrix commented on mkosi/config.py:
```diff
@@ -348,6 +352,28 @@ def to_qemu(self) -> str:
+            Architecture.s390x       : "s390x",
+            Architecture.x86         : "386",
+            Architecture.x86_64      : "amd64",
+        }.get(self)
```
I am unable to check this with all the arches, I just took this list from the specification/documentation

septatrix reviewed (commented):
Please also note the several inline comments I left

behrmann commented on mkosi/config.py:
```diff
@@ -324,6 +326,8 @@ def to_efi(self) -> Optional[str]:
             Architecture.loongarch64 : "loongarch64",
         }.get(self)
 
+        # TODO should this have an explicit check+die?
```
Why the comment here?

septatrix replied:
All the other `to_*` methods have explicit checks if the architecture is supported and exit with a proper error message except for this one. That seemed like an oversight though on second look it seems like the `None` case is used somewhere and explicit checks are in some other places

behrmann commented on mkosi/config.py:
```diff
@@ -157,6 +157,8 @@ class OutputFormat(StrEnum):
     sysext    = enum.auto()
     tar       = enum.auto()
     uki       = enum.auto()
+    # TODO Do we want to support oci-dir or oci-tar, or both?
+    oci       = enum.auto()
```
Go with oci_dir, since that's it and it leaves more open
```suggestion
    # TODO Do we want to support oci-dir or oci-tar, or both?
    oci       = enum.auto()
```

septatrix replied:
I wanted to go with the nomenclature used by the containers tools which use [`oci` and `oci-archive`](https://github.com/containers/image/blob/278eaa3c1b629f0438e784cde703403c44898fc4/docs/containers-transports.5.md#ocipathreference)

septatrix replied:
`docker image import` only seems to allow importing tars so supporting both might be a good idea (at least in the long run).

behrmann replied:
> I wanted to go with the nomenclature used by the containers tools which use [oci and oci-archive](https://github.com/containers/image/blob/278eaa3c1b629f0438e784cde703403c44898fc4/docs/containers-transports.5.md#ocipathreference)

If that's the official nomenclature, then let's go with that. 

> docker image import only seems to allow importing tars so supporting both might be a good idea (at least in the long run).

Is docker still a thing? I thought all the cool kids these days used podman.

septatrix replied:
> If that's the official nomenclature, then let's go with that.

Not by the [opencontainers initiative](https://github.com/opencontainers) itself as they do not specify how image names should be resolved but its the name used by the [containers projects](https://github.com/containers) which actually support working directly with local OCIs.

> Is docker still a thing?

Not on anything I touch but yes, still alive and kicking.

behrmann commented on mkosi/archive.py:
```diff
@@ -63,6 +67,72 @@ def make_tar(src: Path, dst: Path, *, tools: Path = Path("/"), sandbox: Sequence
+
+    layer_archive = context.staging / context.config.output_tar
+    layer_digest = hash_file(layer_archive)
+    # Should we create a symlink at the old path such that the tar can still be used?
```
I don't think we need to generate more output than people ask for, when they want tar, they can generate a tar.

septatrix replied:
I was thinking of matching the current behavior for initrd/efi files which are also provided even when building disk images

behrmann replied:
That's true, but the tar might be quite a bit bigger and we have an explicit tar output. In the end I'm only -0 about it, though.

septatrix replied:
The tar is already created as it is used inside the OCI (also its currently not compressed, see comment further below in the code). So it would only be a symlink to an already existing file

behrmann commented on mkosi/archive.py:
```diff
@@ -63,6 +67,72 @@ def make_tar(src: Path, dst: Path, *, tools: Path = Path("/"), sandbox: Sequence
+    oci_config = {
+        # TODO Add more metadata? Created timestamp, history, config...
+        "architecture": context.config.architecture.to_oci(),
+        "os": "linux",
```
Is that always just "linux" or should that be the distro?

septatrix replied:
That is always linux (https://github.com/opencontainers/image-spec/blob/365fa41472a19107b0b4cbc17e141a9b97f9b66e/config.md?plain=1#L108)

behrmann replied:
What a terrible spec, they could link to whatever they reference. Could you add a comment explaining that this is (I suppose) the first half (i.e. non-arch) output of `go tool dist list`?

septatrix replied:
They do (and it works in the rendered markdown) though I linked the raw markdown where the link target is in the footnotes: https://golang.org/doc/install/source#environment. I will add the comment though

behrmann commented on mkosi/archive.py:
```diff
@@ -63,6 +67,72 @@ def make_tar(src: Path, dst: Path, *, tools: Path = Path("/"), sandbox: Sequence
+    layer_archive.rename(ca_store / layer_digest)
+
+    oci_config = {
+        # TODO Add more metadata? Created timestamp, history, config...
```
timestamp makes sense, though that should probably use `SOURCE_DATE_EPOCH`. Are there any fields that could use the image version or image id? 

No idea what history and config are.

septatrix replied:
> timestamp makes sense, though that should probably use `SOURCE_DATE_EPOCH`.

Good idea.

> Are there any fields that could use the image version or image id?

There is `os.version` though that is for the host system on which this should run (which is also the case for `os`).

Otherwise we have labels/annotations, [many of which are standardized](https://github.com/opencontainers/image-spec/blob/365fa41472a19107b0b4cbc17e141a9b97f9b66e/annotations.md#pre-defined-annotation-keys). For us `org.opencontainers.image.version` would work. There is also `org.opencontainers.image.base.name` but that is actually for base images in the Docker sense, not our base distribution.

> No idea what history and config are.

Config can be used to preset information used when running the image. E.g. an entrypoint/cmd, working dir, environment variables. One idea would be to set `/sbin/init` here though I am not sure if that is the best idea. If we do not set anything at least the entrypoint/cmd always has to be given explicitly. Working dir seems to default to `/` and env is empty apart from what the container runtime sets.
(Users are still able to create a new image building upon the oci output from mkosi and set these afterwards. Though I am not sure about the UX for docker - might be that they have to explicitly load the oci image into the local registry first. With podman this is as simple as `FROM oci:mkosi.output/image\nCMD /usr/bin/bash`)

History is used to log the stuff applied to the image, e.g. commands run, files added etc. Dockerhub seem to show them when viewing details about specific image version, e.g. https://hub.docker.com/layers/library/fedora/40/images/sha256-e7c8351c0292634436207cd10654eff5361153f0617e71c8fbd6e8dcf8d6f826?context=explore. These mostly seem to use the syntax from Dockerfiles though the spec does not set any restrictions on them. I am not sure what would happen if one where to just write random stuff in there.
We could save all invoked commands and store them in the history but that seems rather cumbersome and not worthwhile.

septatrix replied:
> Though I am not sure about the UX for docker - might be that they have to explicitly load the oci image into the local registry first. With podman this is as simple as `FROM oci:mkosi.output/image\nCMD /usr/bin/bash`)

I just checked how this works with docker. Unless I missed something one has to run `tar -c . | docker import - my-mkosi-oci` before being able to use the image. Docker (or buildx) does not support simply using `oci:...` as an image specifier but instead interprets it as an image with name "oci".

behrmann replied:
History and config seem a bit pointless or too far out, but setting `org.opencontainers.image.base.name` might make sense. Wouldn't the images produced by mkosi be for all intents and purpses base images?

septatrix replied:
Yes though the base name annotation is for the other way around. It should refer to the image upon which this image is based. E.g. if I create an image based on `registry.fedoraproject.org/fedora:latest` then that would saved in the annotations.

septatrix replied:
I tried out setting a `Cmd` in config such that `/sbin/init` is started automatically, I think that might be a reasonable thing to do but we can also drop it. I am a bit toying around with what would make sense.

I also added a single history field for testing with a `comment` and `created_by` field. The comment does not seem to show up but the created_by field is at least shown in Docker Hub for my demo image: https://hub.docker.com/layers/septatrix/mkosi-oci-fedora-test/latest/images/sha256-a81676033c96a8de79a3441640fbfb3ca14da09276e1409281e7ce02ecc44e55?context=explore

septatrix commented on mkosi/util.py:
```diff
@@ -258,3 +259,15 @@ def _write_contents(target, source):
+        while (buf := sf.read(bs)):
+            h.update(buf)
+    
+    return h.hexdigest()
```
Oh no, I have become the very thing I swore to destroy ^^ (While working I started to use the shortcut for save without formatting which also disables adding the trailing newline, after a while I simply disabled Black for this repo but that did not save this normally I guess).

septatrix commented:
Should I add something like podman in the CI to test the OCI? A runtime-spec for `systemd-nspawn --oci-bundle` should also be possible but that would be a bit more complex

behrmann commented:
I guess. If we are to add this, it should be covered by a test.

septatrix commented:
Thinking about the tests: Unpacking the OCI bundle and creating the runtime spec is rather annoying and [`umoci`](https://github.com/opencontainers/umoci) which would be great for this is not package in most places. So while using `sd-nspawn --oci-bundle` would be neat it is rather annoying and unpacking the OCI and creating the runtime spec on our own would add places where we could hide our own bugs (apart from being annoying).

As a result I think using podman to test this is likely the simplest way for now. However, installing podman in the tools tree also has it problems as the tools tree is inaccessible from within the pytest tests as the proper config parsing etc is only done within mkosi itself. So using the tools tree version would only be possible if we add podman support to the boot/shell verbs. I am not sure if this is desired as this would also likely behave differently regarding stuff like `RuntimeTree`, network namespaces, etc...

Alternatively we could install podman outside of the tools tree.

Which variant do you prefer?

septatrix commented:
I have added a basic tests using podman

septatrix reviewed (commented):
The CI failures for Arch and Debian testing seem to be due to https://github.com/systemd/systemd/issues/29860. No clue why CentOS fails and I did not find the motivation to lock through all the logs yet.

Regardless: This is not a problem with the way the OCI images are constructed but a bug/misconfiguration in either podman or systemd. Otherwise all CI runs would fail. I am actually inclined to add skips for the failing scenarios. People can still use the OCI generated by mkosi and start the tools directly instead of going via the init system. Merging this functionality in mkosi could even be beneficial for podman/systemd to better diagnose who exactly is at fault here.

septatrix commented:
> No clue why CentOS fails and I did not find the motivation to lock through all the logs yet.

It's an endless loop due to journald not starting properly (which is also why the timeout after 1h kicked in. The repeating log section is attached. I have found a few systemd issues which seem to reference what I would guess is the error cause (`PR_SET_MM_ARG_START failed...`). As this does not happen in the other distros I would assume that this has since been fixed in upstream systemd further making me lean towards skipping the OCI testing for centos.

<details>
<summary>Log extract</summary>

```txt
         Starting Journal Service...
systemd-tmpfiles-setup.service: starting held back, waiting for: systemd-journald.service
systemd-journald.service: Failed to add invocation ID to keyring, ignoring: Operation not permitted
Operating on architecture: x86
Operating on architecture: x32
Operating on architecture: x86-64
Operating on architecture: x86
Operating on architecture: x32
Operating on architecture: x86-64
Operating on architecture: x86
Operating on architecture: x32
Operating on architecture: x86-64
Restricting namespace to: n/a.
Operating on architecture: x86
Blocking cgroup.
Blocking ipc.
Blocking net.
Blocking mnt.
Blocking pid.
Blocking user.
Blocking uts.
Blocking time.
Operating on architecture: x32
Blocking cgroup.
Blocking ipc.
Blocking net.
Blocking mnt.
Blocking pid.
Blocking user.
Blocking uts.
Blocking time.
Operating on architecture: x86-64
Blocking cgroup.
Blocking ipc.
Blocking net.
Blocking mnt.
Blocking pid.
Blocking user.
Blocking uts.
Blocking time.
Operating on architecture: x86
Operating on architecture: x32
Operating on architecture: x86-64
Operating on architecture: x86-64
[**    ] (2 of 2) A start job is running for…odes in /dev (6min 41s / no limit)
M
[K[*     ] (2 of 2) A start job is running for…odes in /dev (6min 42s / no limit)
M
[K[**    ] (2 of 2) A start job is running for…odes in /dev (6min 42s / no limit)
M
[K[***   ] (1 of 2) A start job is running for Journal Service (3s / no limit)
M
[K[ ***  ] (1 of 2) A start job is running for Journal Service (4s / no limit)
M
[K[  *** ] (1 of 2) A start job is running for Journal Service (4s / no limit)
M
[K[   ***] (2 of 2) A start job is running for…odes in /dev (6min 44s / no limit)
M
[K[    **] (2 of 2) A start job is running for…odes in /dev (6min 45s / no limit)
M
[K[     *] (2 of 2) A start job is running for…odes in /dev (6min 45s / no limit)
M
[K[    **] (1 of 2) A start job is running for Journal Service (6s / no limit)
M
[K[   ***] (1 of 2) A start job is running for Journal Service (7s / no limit)
M
[K[  *** ] (1 of 2) A start job is running for Journal Service (7s / no limit)
M
[K[ ***  ] (2 of 2) A start job is running for…odes in /dev (6min 47s / no limit)
M
[K[***   ] (2 of 2) A start job is running for…odes in /dev (6min 48s / no limit)
M
[K[**    ] (2 of 2) A start job is running for…odes in /dev (6min 48s / no limit)
M
[K[*     ] (1 of 2) A start job is running for Journal Service (9s / no limit)
M
[K[**    ] (1 of 2) A start job is running for Journal Service (10s / no limit)
Received SIGCHLD from PID 64 (systemd-journal).
Child 64 (systemd-journal) died (code=exited, status=1/FAILURE)
systemd-journald.service: Child 64 belongs to systemd-journald.service.
systemd-journald.service: Main process exited, code=exited, status=1/FAILURE
systemd-journald.service: Failed with result 'exit-code'.
systemd-journald.service: Service will restart (restart setting)
systemd-journald.service: Changed start -> failed
systemd-journald.service: Job 293 systemd-journald.service/start finished, result=failed
M
[K[FAILED] Failed to start Journal Service.
[KSee 'systemctl status systemd-journald.service' for details.
systemd-journald.socket: Changed running -> listening
systemd-journald.service: Unit entered failed state.
systemd-journald.service: Consumed 15ms CPU time.
systemd-journald.service: Changed failed -> auto-restart
systemd-journald.service: Control group is empty.
systemd-journald.service: Service has no hold-off time (RestartSec=0), scheduling restart.
systemd-journald.service: Trying to enqueue job systemd-journald.service/restart/replace
systemd-journald.service: Installed new job systemd-journald.service/restart as 298
systemd-journald.service: Enqueued job systemd-journald.service/restart as 298
systemd-journald.service: Scheduled restart job, restart counter is at 42.
systemd-journald.socket: Incoming traffic
systemd-journald.socket: Changed listening -> running
sysinit.target: starting held back, waiting for: systemd-resolved.service
systemd-journal-flush.service: starting held back, waiting for: systemd-journald.service
systemd-journald.service: Changed auto-restart -> dead
systemd-journald.service: Job 298 systemd-journald.service/restart finished, result=done
[  OK  ] Stopped Journal Service.
systemd-journald.service: Converting job systemd-journald.service/restart -> systemd-journald.service/start
systemd-journald.service: Consumed 15ms CPU time.
sysinit.target: starting held back, waiting for: systemd-resolved.service
systemd-journal-flush.service: starting held back, waiting for: systemd-journald.service
systemd-journald.service: Will spawn child (service_enter_start): /usr/lib/systemd/systemd-journald
systemd-journald.service: Failed to set 'trusted.invocation_id' xattr on control group /system.slice/systemd-journald.service, ignoring: Operation not permitted
systemd-journald.service: Failed to remove 'trusted.delegate' xattr flag on control group /system.slice/systemd-journald.service, ignoring: Operation not permitted
systemd-journald.service: Passing 3 fds to service
systemd-journald.service: About to execute /usr/lib/systemd/systemd-journald
systemd-journald.service: Forked /usr/lib/systemd/systemd-journald as 65
PR_SET_MM_ARG_START failed: Operation not permitted
systemd-journald.service: Changed dead -> start
```

</details>

DaanDeMeyer commented:
We're trying to get `umoci` packaged in Fedora so then it should be OK to depend on it.

DaanDeMeyer commented:
@septatrix Looking at https://pkgs.org/download/umoci, it seems that only centos/fedora are missing and the rest is covered already. Are there any particular distributions you have in mind?

septatrix commented:
Oh in that case I guess not. I was looking at their readme but I guess that's quite outdated: https://github.com/opencontainers/umoci?tab=readme-ov-file#install

septatrix commented on tests/test_boot.py:
```diff
@@ -40,6 +40,28 @@ def test_boot(config: Image.Config, format: OutputFormat) -> None:
+            # or systemd resolved this upstream.
+            # TODO it seems that mounting an empty file at /etc/machine-id also works
+            # but than systemd-machine-id-commit.service fails as /etc/ is not on a temporary file system...
+            # And that would also skip the first-boot logic.
```
I marked these for skipping as I do not consider supporting containers with systemd inside them to be mandatory for this feature and it is unlikely that there is something we could do to improve this.

However, I am tempted to /cc someone like rhatdan who might know more about the podman side of things. Do you perhaps know anyone more suited for this before I ping him?

septatrix replied:
@rhatdan Are you aware of any changes in this area? It seems like the environment which podman provides is no longer sufficient for systemd to run. Should I create an issue for this in the podman repo?

septatrix commented on mkosi/resources/mkosi.md:
```diff
@@ -666,11 +667,12 @@ boolean argument: either `1`, `yes`, or `true` to enable, or `0`, `no`,
   either a boolean or a compression algorithm (`xz`, `zstd`). `zstd`
   compression is used by default, except CentOS and derivatives up to version
-  8, which default to `xz`. Note that when applied to block device image types,
+  8, which default to `xz`, and OCI images, which default to `gzip`.
```
Before anyone asks, OC**I** **i**mages is correct, as the first abbreviation stands for open container initiative

septatrix commented:
Is there anything else which has to be addressed? The failing test was flaky at the time of my last push

keszybz commented on mkosi/config.py:
```diff
@@ -200,6 +202,18 @@ def extension(self) -> str:
 
+    def oci_media_type_suffix(self) -> str:
+        suffix = {
+            Compression.none: "",
+            Compression.gz: "+gzip",
+            Compression.zstd: "+zstd",
```
Please align the strings to a column.

keszybz commented on mkosi/__init__.py:
```diff
@@ -2806,6 +2796,104 @@ def make_disk(
+    maybe_compress(
+        context,
+        context.config.compress_output,
+        context.staging / "rootfs.layer",
+        context.staging / "rootfs.layer",
```
If the same arg is given twice on purpose, maybe add a brief comment what it happenning here.

keszybz commented on mkosi/__init__.py:
```diff
@@ -2806,6 +2796,104 @@ def make_disk(
+    oci_config = {
+        "created": creation_time,
+        "architecture": context.config.architecture.to_oci(),
+        # Name of the operating system which the image is built to run on.
+        # SHOULD use value listed in the golang documentation for $GOOS.
+        # https://github.com/opencontainers/image-spec/blob/v1.0.2/config.md#properties
```
I was confused by this comment. "SHOULD" makes it sound like something is not implemented here. Maybe just say "Name of the operating system which the image is built to run on as defined by https://github.com/opencontainers/image-spec/blob/v1.0.2/config.md#properties."

keszybz commented:
> We're trying to get `umoci` packaged in Fedora so then it should be OK to depend on it.

Link?

keszybz commented on tests/test_boot.py:
```diff
@@ -40,6 +40,31 @@ def test_boot(config: Image.Config, format: OutputFormat) -> None:
+
+            # These services fail as /etc/machine-id cannot be properly initialized.
+            # This is not possible as overmounting is not permitted inside the container.
+            # Running the container with higher privileges using `--privileged` or `--cap-add SYS_ADMIN`
```
Please don't use backticks. Backticks only make sense in markdown, and this is not markdown.

DaanDeMeyer commented:
> > We're trying to get `umoci` packaged in Fedora so then it should be OK to depend on it.
> 
> Link?

No link, Michel Lind has been working on it.

I don't think we want to merge this until we figure out whether we can test this with systemd-nspawn instead of podman. I'd much rather get more coverage on systemd-nspawn's OCI support than having to figure out various podman failures when these tests inevitably start failing.

septatrix commented:
> I don't think we want to merge this until we figure out whether we can test this with systemd-nspawn instead of podman. I'd much rather get more coverage on systemd-nspawn's OCI support than having to figure out various podman failures when these tests inevitably start failing.

Another issue with using umoci would be that it only generates the minimal OCI runtime bundle and [has no means to specify e.g. mounts](https://github.com/opencontainers/umoci/issues/504). The means we would have to set up [all the stuff required by systemd](https://systemd.io/CONTAINER_INTERFACE/) ourselves, likely by modifying the JSON spec generated by umoci. This means that we introduce a new step where we could hide existing errors or add our own ones

DaanDeMeyer commented on mkosi/__init__.py:
```diff
@@ -3090,6 +3178,20 @@ def build_image(context: Context) -> None:
+            context.root, context.staging / "rootfs.layer",
+            tools=context.config.tools(),
+            sandbox=context.sandbox,
+            # We do not want SELinux xattrs in OCI layers
+            # as they seldom make sense inside containers.
+            options=["--xattrs-exclude", "security.selinux"],
```
People make VMs now from OCI images, so I think excluding selinux is a bit presumptuous. I would prefer the selinux xattrs are stripped out when extracting the OCI image, not when creating it.

septatrix replied:
umoci logged warnings when extracting an OCI with selinux labels but I can keep them in if that is preferred

DaanDeMeyer reviewed (changes requested):
Can we drop the stuff for running OCI images with `podman`? I really don't think we should add that. I'm OK with the OCI output format though.

Also please squash commits, reviewing unsquashed commits is just pure pain.

septatrix commented:
I rebased the branch and squashed some commit. I hope the split are done in a way which makes review easier.

I also dropped the podman stuff. This means that only building is tested and not running but until we find a better solution (e.g. umoci, sd-nspawn with a manually created OCI bundle, lxc import with oci-template) this should be fine. We can still revisit this at a later time (e.g. when umoci is included everywhere and then include it in the tools tree).

DaanDeMeyer commented on mkosi/resources/mkosi.md:
```diff
@@ -705,11 +706,12 @@ boolean argument: either `1`, `yes`, or `true` to enable, or `0`, `no`,
   either a boolean or a compression algorithm (`xz`, `zstd`). `zstd`
   compression is used by default, except CentOS and derivatives up to version
-  8, which default to `xz`. Note that when applied to block device image types,
+  8, which default to `xz`, and OCI images, which default to `gzip`.
```
Why default to gzip for OCI images? We generally don't compress by default, I wouldn't do so for OCI images either.

DaanDeMeyer commented on mkosi/__init__.py:
```diff
@@ -3261,7 +3261,99 @@ def make_disk(
 def make_oci(context: Context, root_layer: Path, dst: Path) -> None:
-    raise NotImplementedError
+    ca_store = dst / "blobs" / "sha256"
+    ca_store.mkdir(parents=True)
```
Use `with umask` here to make sure the mode is correct. Should be good with 755 here.

septatrix replied:
Do you mean `os.umask` or the mode parameter of mkdir?

DaanDeMeyer commented on mkosi/__init__.py:
```diff
@@ -3260,6 +3260,10 @@ def make_disk(
     return make_image(context, msg=msg, skip=skip, split=split, tabs=tabs, root=context.root, definitions=definitions)
 
 
+def make_oci(context: Context, root_layer: Path, dst: Path) -> None:
```
Just name `root_layer` `root` or so

DaanDeMeyer commented on mkosi/__init__.py:
```diff
@@ -3261,7 +3261,99 @@ def make_disk(
+    oci_manifest_digest = hashlib.sha256(oci_manifest_blob.encode()).hexdigest()
+    (ca_store / oci_manifest_digest).write_text(oci_manifest_blob)
+
+    with (dst / "index.json").open("w") as f:
```
Explicit mode here please

DaanDeMeyer commented on mkosi/__init__.py:
```diff
@@ -3261,7 +3261,99 @@ def make_disk(
+    }
+    oci_manifest_blob = json.dumps(oci_manifest)
+    oci_manifest_digest = hashlib.sha256(oci_manifest_blob.encode()).hexdigest()
+    (ca_store / oci_manifest_digest).write_text(oci_manifest_blob)
```
`with umask` here to make sure the mode is correct.

DaanDeMeyer commented on mkosi/__init__.py:
```diff
@@ -3261,7 +3261,99 @@ def make_disk(
+            f,
+        )
+
+    with (dst / "oci-layout").open("w") as f:
```
Explicit mode

DaanDeMeyer commented on mkosi/__init__.py:
```diff
@@ -3261,7 +3261,99 @@ def make_disk(
+    oci_manifest_digest = hashlib.sha256(oci_manifest_blob.encode()).hexdigest()
+    (ca_store / oci_manifest_digest).write_text(oci_manifest_blob)
+
+    with (dst / "index.json").open("w") as f:
+        json.dump(
```
Just use `json.dumps` + `Path.write_text()` here please

DaanDeMeyer commented on mkosi/__init__.py:
```diff
@@ -3261,7 +3261,99 @@ def make_disk(
+            f,
+        )
+
+    with (dst / "oci-layout").open("w") as f:
+        json.dump({"imageLayoutVersion": "1.0.0"}, f)
```
`Path.write_text()` here as well please

DaanDeMeyer reviewed (changes requested):
Looks great, but please make sure all output files and directories are created with explicit permissions. Also, please merge the second and third commit.

DaanDeMeyer commented:
@septatrix This would be a separate PR, but given the "layers" stuff in OCI images, it'd be cool if we could integrate this properly with `Overlay=yes` and `BaseTrees=` somehow.

DaanDeMeyer commented on mkosi/__init__.py:
```diff
@@ -3270,6 +3260,104 @@ def make_disk(
+            "Cmd": [
+                "/sbin/init",
+                *context.config.kernel_command_line,
+                *context.config.kernel_command_line_extra,
```
Oops, I missed this, please drop this, KernelCommandLineExtra= is for extra stuff to be passed at runtime, not when building the image.

DaanDeMeyer replied:
Fixed this myself

behrmann commented on mkosi/__init__.py:
```diff
@@ -3270,6 +3260,103 @@ def make_disk(
+
+    creation_time = (
+        datetime.datetime.fromtimestamp(context.config.source_date_epoch, tz=datetime.timezone.utc)
+        if context.config.source_date_epoch
```
You probably want to check `if context.config.source_date_epoch is not None` here, since otherwise the timestamp 0 is illegal and would lead to the `now` branch.

septatrix commented on mkosi/config.py:
```diff
@@ -634,6 +670,8 @@ def config_default_compression(namespace: argparse.Namespace) -> Compression:
         else:
             return Compression.zstd
+    elif namespace.output_format == OutputFormat.oci:
+        return Compression.gzip
```
Zstd support is not yet required by the spec: https://github.com/opencontainers/image-spec/issues/803

cyphar commented:
It's a bit of a shame I didn't see this thread back when you were discussing using umoci -- I would've been quite happy to help! :smile_cat:

umoci is packaged for most distributions, and it is an optional requirement for LXC and Incus (as well as for KIWI but that's something primarily used by (open)SUSE AFAIK) so anything that packages those probably also packages umoci. It's a very easy program to package (speaking as a package maintainer for openSUSE) so I wouldn't expect that to be a large hurdle to overcome. I guess I should add a reference to https://pkgs.org/download/umoci or similar to the README.

As for the discussion about setting runtime-spec `config.json` parameters -- I would like to know more about what in particular you would like to configure. Most of the things in the [systemd container interface specification](https://systemd.io/CONTAINER_INTERFACE/) are handled by OCI container runtimes and do not require explicit configuration AFAICS.

My original goal was that `umoci unpack` would give you a basic runtime configuration with everything from the image itself populated (such as resource limits, cmdline options, etc etc) and then users could add whatever extra bits they require to provide as much flexibility as possible. Notably [I wrote the specification for converting from image-spec configurations to runtime-spec configurations](https://github.com/opencontainers/image-spec/blob/main/conversion.md).

I felt that adding a bunch of extra configuration knobs would be out of place for a tool for images -- the runtime-spec is quite trivial to modify using `jq` or any other general-purpose JSON library (if you prefer a purpose-built-tool then there's [`oci-runtime-tools`](https://github.com/opencontainers/runtime-tools) but I would personally recommend modifying the JSON directly if what you need is trivial). There is a very large number of possible things that a user might want to configure that it would be unreasonable to add a knob for every one to umoci.

That being said, I am open to discussing what kinds of knobs you might need.
